### ISSUE TITLE
Training details with mobilenet V2 on Imagenet

### ISSUE THREAD

**Issue by zs271** [cmu-enyac/LeGR#2]
Hi Rudy,

Can you provide training details (or scripts) for prunning mobilenet V2 on imagenet (e.g., number of GPUs, batch size, tau_hat, pruning ratio when searching for the transofrmation, lr, lr decay policy, weight decay, etc.). Many thanks.

Best wishes,
Chen

**Comment by RudyChin** [comment]
Hi Chen,

Thank you for your interests in our work! For MobileNetV2, we train the networks on TPU using [this repository](https://github.com/RudyChin/LeGR-MobileNetV2-TPU). The hyper-params are in the readme there.

Thanks,
Rudy

**Comment by zs271** [comment]
Thanks. Great work!

BTW, it seems quite difficult to get to the target accuracy when training on mobilenet V2 (when pruned to 70% FLOPS I only got 69.8%). How many epochs did you train and are there tricks to get to the target 71.4% accuracy? 

Many thanks.

**Comment by RudyChin** [comment]
Hi Chen,

Yes, I think it is trickier to train MobileNetV2 compared to training ResNet. I used the hyper-parameters provided by MnasNet on TPU to train it, which is in the repository that I mentioned above. You can find all the hyper-params there. Hope this helps.

Thanks,
Rudy

**Comment by zs271** [comment]
Hi Rudy,

Thanks, MobileNet V2 is indeed quite difficult to train and requires many epochs for fine tuning. Can you also provide codes for pruning ResNet 50 on Imagenet? Many thanks.

Best,
Chen

**Comment by RudyChin** [comment]
Hi Chen,

Sorry for the "really late" reply...
For ImageNet, LeGR does the same thing, so the code is the same. The only thing that is different is the data we used during fine-tuning. The data augmentation for ImageNet is in https://github.com/cmu-enyac/LeGR/blob/master/utils/drivers.py#L67
So, all you need a pre-trained ResNet-50 model from torchvision to use LeGR.

Thanks,
Rudy

**Comment by WeiyiLi** [comment]
@RudyChin Hello Rudy, could you please also provide the training details for training ResNet50 on ImageNet? (e.g., batch size, tau_hat, pruning ratio when searching for the transformation, lr, lr decay policy, weight decay, etc.) Thank you very much!

**Comment by RudyChin** [comment]
Hi Weiyi,

Thank you for your interest in our work! For searching the layer-wise affine transformations, I've done the search using a single GPU (NVIDIA 1080 Ti) and the hyperparameters have not been tuned, it is possible that one can improve our results by searching these hyperparameters.
- Batch size 96
- tau_hat 200
- pruning down to 47% of the original FLOPs (the smallest considered in the paper)
- lr 1e-2 (constant for the entire 200 iterations)
- optimizer: SGD w/ momentum, weight decay 5e-4

Once the affine transformations are found, fine-tuning is done using the following hyperparameters:
- 60 epochs
- batch size 192 (due to memory limitation)
- lr 1e-2, decay by 10x at epoch 30 and 45
- optimizer: SGD w/ momentum, weight decay 5e-4

Hope this helps!

Best,
Rudy

**Comment by WeiyiLi** [comment]
@RudyChin Thank you very much for these information. I will try it out and hopefully I can get the results as good as you do. :)

**Comment by WeiyiLi** [comment]
Hi @RudyChin , is it possible to use this code to train Mobilenetv2 on ImageNet? If yes, could you please provide the link of the pretrained model? I tried to load the network using your MobileNetV2.py and pytorch's pretrained weight but error is reported, so I guess maybe I shouldn't use the pytorch's pretrained weight...

**Comment by WeiyiLi** [comment]
@RudyChin Thanks for your support on ResNet50 and the result looks good. Then I tried MobileNetV2 on ImageNet. I changed the code a little bit to make support MobileNetV2 trained in https://github.com/d-li14/mobilenetv2.pytorch . The parameters I used are

- Batch size 96
- tau_hat 200
- pruning down to 50% as your paper suggested
- lr 1e-2 (constant for the entire 200 iterations) 
- optimizer: SGD w/ momentum, weight decay 5e-4

After  obtaining the affine transformation, the parameters I used in fine-tuning are 
- 60 epochs
- batch size 96
- lr 1e-2, decay by 5x at epoch 18, 36 and 45 (as your original code did)
- optimizer: SGD w/ momentum, weight decay 5e-4

However, after fine-tuning, I can only get the top1 accuracy 64.28%, whose original top1 is 72.79%. (In your paper, you can get more than 69%)
Could you please give any suggestions, such as the hyperparameters to reach the accuracy as your paper stated? Should I decrease my learning rate more and run more epochs? Or can you provide your hyperparameters as reference? Thank you so much.

**Comment by RudyChin** [comment]
Hi Weiyi,

As for MobileNetV2, I use the same setting to find affine-transformation but use the TPU code to train MobileNetV2.
The repo for that is https://github.com/RudyChin/LeGR-MobileNetV2-TPU where the hyperparameters are detailed in that repo.
I find hyperparams to be very important for mobilenets and instead of tuning those, I simply train the found architecture using the TPU code and follow their hyperparams.

Hope this helps!
